### DISCUSSION TITLE
A2DPBluetoothSource Connecting / Pairing

### DISCUSSION THREAD

**Discussion by Prog32aw** [pschatzmann/arduino-audio-tools#1781] [Q&A] [answered]
I use 5 different Bluetooth sound devices, that all work without problems with a smartphone or pc as a source. But only 1 can be used with the A2DPBluetoothSource. Any idea how to connect them?
All devices has been set to pairing mode before.

Thanks!

**Comment by pschatzmann** [comment, marked as answer]
Just activate the logging and compare the logs.
You might need to ajust the supported cod service if your speaker reports something very exotic...

**Reply by Prog32aw** [reply]
I did compare and found a problem in the bluetooth device´s name. The full name is "Bose Mini SoundLink" (is it shown on other devices with this name) and can only be connected as "Bose Mini Sou", as it can be found in the log output.  Connection and sound output could be established. Furthermore auto_reconnect=true seems to prevent to connect to new devices.

The BluetoothA2DPSource.begin can be called with a vector of names. Is it correct that the device to be connected must be always in pairing mode, even when it was connected before? 
The only connection I could establish with devices not in pairing mode was with cfgA2DP.auto_reconnect set to "true", when it was connected in pairing mode before.
Setting the devices in pairing mode before connecting (when already paired) would be a litte bit inconvenient.... 

Thanks

**Comment by pschatzmann** [comment]
I am not aware of any such requirement: at least my bluetooth speaker would not have this!
The only restriction that I fund is, that you might not be able to reconnect to the bluettooth speaker sometimes when the connection was not closed properly and the speaker thinks that it is still connected.

But read the __documentaton from your Bluetooth Speaker__ to find the correspnding details.


I never recommended to use auto_reconnect = true and it does what it is documented to do: it tries to reconnect to the last device for the specified time period.

**Reply by Prog32aw** [reply]
Thanks for your quick reply, which I do not understand. Is it neccessary to put the Bluetooth device in pairing mode everytime i want to connect?
I dont mean any special condition. Both devices are off before. Do they connect if the sink is not in pairing mode?

**Reply by pschatzmann** [reply]
As I was writing above, for my Bluetooth Speaker this just works and I just need to switch it on.
But yours might work differently, so check it's documentation! 

I guess you can easily test this with your phone: if you switch it on and and you see your speaker appearing, then your ESP32 should also be able to connect to it

**Comment by Prog32aw** [comment]
> But read the documentaton from your Bluetooth Speaker to find the correspnding details.

This is always the same procedure, as you probably know: put the device in pairing mode, search it, connect to its name (address), and thats all. I think the Source should maintain a list of sink device names and addresses and tries to connect them by address on each start.

With BluetoothA2DP I experienced, that the device must always be in pairing mode to connected again, except when auto_reconnect is true. A look into the code shows that the last address is stored to NVS.

**Comment by Prog32aw** [comment]
> As I was writing above, for my Bluetooth Speaker this just works and I just need to switch it on.
But yours might work differently, so check it's documentation!

Maybe your speaker tries to establish the connection by itself?

> I guess you can easily test this with your phone: if you switch it on and and you see your speaker appearing, then your ESP32 should also be able to connect to it

All devices I have behave the same: they are always reconnected by a phone, pc, tablet what ever automatically (not in pairing mode), when they have been paired before.
None of them do this with BluetoothA2DPSource. 
Only one device will do when auto_reconnect = true.

**Comment by pschatzmann** [comment]
Can you share the log (with the core debug level set to info and with autoreconnect = false)? 
Start the Bluetooth Speaker first
Then Start the ESP32

ps. also make sure that your phone is not automatically connecting and blocking the ESP32 from working

**Comment by Prog32aw** [comment]
auto_reconnect = false:
[   132][D][BluetoothA2DPSource.cpp:153] start(): [BT_API] start, 
[   132][D][BluetoothA2DPSource.cpp:172] start_raw(): [BT_API] start_raw, 
[   135][D][BluetoothA2DPCommon.cpp:163] get_last_connection(): [BT_AV] get_last_connection
[   144][D][BluetoothA2DPCommon.cpp:186] get_last_connection(): [BT_AV] => 04:5d:4b:e9:43:4c
[   655][D][BluetoothA2DPSource.cpp:326] bt_app_task_start_up(): [BT_AV] bt_app_task_start_up
[   656][D][BluetoothA2DPSource.cpp:257] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0x0, param len 0
[   663][D][BluetoothA2DPSource.cpp:307] bt_app_task_handler(): [BT_API] bt_app_task_handler, signal: 0x1, event: 0x0
[   674][D][BluetoothA2DPSource.cpp:563] bt_av_hdl_stack_evt(): [BT_AV] bt_av_hdl_stack_evt event: 0
[   683][I][BluetoothA2DPCommon.cpp:358] set_scan_mode_connectable(): [BT_AV] set_scan_mode_connectable false
[   692][I][BluetoothA2DPSource.cpp:600] bt_av_hdl_stack_evt(): [BT_AV] Starting device discovery...
[   695][D][BluetoothA2DPSource.cpp:257] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0x4, param len 16
[   712][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 10
[   712][D][BluetoothA2DPSource.cpp:307] bt_app_task_handler(): [BT_API] bt_app_task_handler, signal: 0x1, event: 0x4
[   722][I][BluetoothA2DPSource.cpp:555] bt_app_gap_callback(): [BT_AV] event: 10
[   732][I][BluetoothA2DPSource.cpp:665] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state APP_AV_STATE_DISCOVERING, evt 0x4
[   739][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 10
[   750][D][BluetoothA2DPSource.cpp:632] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks
[   759][I][BluetoothA2DPSource.cpp:555] bt_app_gap_callback(): [BT_AV] event: 10
[   775][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 10
[   786][I][BluetoothA2DPSource.cpp:555] bt_app_gap_callback(): [BT_AV] event: 10
[   792][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 10
[   800][I][BluetoothA2DPSource.cpp:555] bt_app_gap_callback(): [BT_AV] event: 10
[   812][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 1
[   816][I][BluetoothA2DPSource.cpp:484] bt_app_gap_callback(): [BT_AV] Discovery started.
[ 10700][D][BluetoothA2DPSource.cpp:257] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0xff00, param len 0
[ 10701][D][BluetoothA2DPSource.cpp:307] bt_app_task_handler(): [BT_API] bt_app_task_handler, signal: 0x1, event: 0xff00
[ 10711][I][BluetoothA2DPSource.cpp:665] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state APP_AV_STATE_DISCOVERING, evt 0xff00
[ 10722][D][BluetoothA2DPSource.cpp:632] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks
[ 13616][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 1
[ 13617][I][BluetoothA2DPSource.cpp:478] bt_app_gap_callback(): [BT_AV] Device discovery failed, continue to discover...
[ 13629][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 1
[ 13633][I][BluetoothA2DPSource.cpp:484] bt_app_gap_callback(): [BT_AV] Discovery started.
[ 20700][D][BluetoothA2DPSource.cpp:257] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0xff00, param len 0
[ 20701][D][BluetoothA2DPSource.cpp:307] bt_app_task_handler(): [BT_API] bt_app_task_handler, signal: 0x1, event: 0xff00
[ 20711][I][BluetoothA2DPSource.cpp:665] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state APP_AV_STATE_DISCOVERING, evt 0xff00
[ 20722][D][BluetoothA2DPSource.cpp:632] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks
[ 26433][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 1
[ 26434][I][BluetoothA2DPSource.cpp:478] bt_app_gap_callback(): [BT_AV] Device discovery failed, continue to discover...
[ 26446][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 1
[ 26450][I][BluetoothA2DPSource.cpp:484] bt_app_gap_callback(): [BT_AV] Discovery started.
[ 30700][D][BluetoothA2DPSource.cpp:257] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0xff00, param len 0
[ 30701][D][BluetoothA2DPSource.cpp:307] bt_app_task_handler(): [BT_API] bt_app_task_handler, signal: 0x1, event: 0xff00
[ 30711][I][BluetoothA2DPSource.cpp:665] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state APP_AV_STATE_DISCOVERING, evt 0xff00
[ 30722][D][BluetoothA2DPSource.cpp:632] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks
[ 39251][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 1
[ 39252][I][BluetoothA2DPSource.cpp:478] bt_app_gap_callback(): [BT_AV] Device discovery failed, continue to discover...
[ 39265][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 1
[ 39269][I][BluetoothA2DPSource.cpp:484] bt_app_gap_callback(): [BT_AV] Discovery started.
[ 40700][D][BluetoothA2DPSource.cpp:257] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0xff00, param len 0
[ 40701][D][BluetoothA2DPSource.cpp:307] bt_app_task_handler(): [BT_API] bt_app_task_handler, signal: 0x1, event: 0xff00
[ 40711][I][BluetoothA2DPSource.cpp:665] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state APP_AV_STATE_DISCOVERING, evt 0xff00
[ 40722][D][BluetoothA2DPSource.cpp:632] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks




auto_reconnect = true:
[   132][D][BluetoothA2DPSource.cpp:153] start(): [BT_API] start, 
[   132][D][BluetoothA2DPSource.cpp:172] start_raw(): [BT_API] start_raw, 
[   135][D][BluetoothA2DPCommon.cpp:163] get_last_connection(): [BT_AV] get_last_connection
[   144][D][BluetoothA2DPCommon.cpp:186] get_last_connection(): [BT_AV] => 04:5d:4b:e9:43:4c
[   642][D][BluetoothA2DPSource.cpp:326] bt_app_task_start_up(): [BT_AV] bt_app_task_start_up
[   643][D][BluetoothA2DPSource.cpp:257] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0x0, param len 0
[   651][D][BluetoothA2DPSource.cpp:307] bt_app_task_handler(): [BT_API] bt_app_task_handler, signal: 0x1, event: 0x0
[   661][D][BluetoothA2DPSource.cpp:563] bt_av_hdl_stack_evt(): [BT_AV] bt_av_hdl_stack_evt event: 0
[   670][I][BluetoothA2DPCommon.cpp:358] set_scan_mode_connectable(): [BT_AV] set_scan_mode_connectable false
[   679][W][BluetoothA2DPSource.cpp:594] bt_av_hdl_stack_evt(): [BT_AV] Reconnecting to 04:5d:4b:e9:43:4c
[   683][D][BluetoothA2DPSource.cpp:257] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0x4, param len 16
[   689][W][BluetoothA2DPCommon.cpp:45] connect_to(): [BT_AV] connect_to to 04:5d:4b:e9:43:4c
[   699][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 10
[   707][I][BluetoothA2DPCommon.cpp:358] set_scan_mode_connectable(): [BT_AV] set_scan_mode_connectable false
[   716][I][BluetoothA2DPSource.cpp:555] bt_app_gap_callback(): [BT_AV] event: 10
[   726][I][BluetoothA2DPSource.h:291] esp_a2d_connect(): [BT_AV] a2dp connecting to: 04:5d:4b:e9:43:4c
[   733][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 10
[   742][D][BluetoothA2DPSource.cpp:307] bt_app_task_handler(): [BT_API] bt_app_task_handler, signal: 0x1, event: 0x4
[   751][I][BluetoothA2DPSource.cpp:555] bt_app_gap_callback(): [BT_AV] event: 10
[   761][I][BluetoothA2DPSource.cpp:665] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state APP_AV_STATE_CONNECTING, evt 0x4
[   769][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 10
[   779][D][BluetoothA2DPSource.cpp:632] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks
[   788][I][BluetoothA2DPSource.cpp:555] bt_app_gap_callback(): [BT_AV] event: 10
[   798][D][BluetoothA2DPSource.cpp:722] bt_app_av_state_connecting_hdlr(): [BT_AV] bt_app_av_state_connecting_hdlr evt 4
[   805][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 10
[   815][W][BluetoothA2DPSource.cpp:758] bt_app_av_state_connecting_hdlr(): [BT_AV] bt_app_av_state_connecting_hdlr unhandled event: 4
[   824][I][BluetoothA2DPSource.cpp:555] bt_app_gap_callback(): [BT_AV] event: 10
[   853][D][BluetoothA2DPSource.cpp:257] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0x0, param len 16
[   855][D][BluetoothA2DPSource.cpp:307] bt_app_task_handler(): [BT_API] bt_app_task_handler, signal: 0x1, event: 0x0
[   865][I][BluetoothA2DPSource.cpp:665] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state APP_AV_STATE_CONNECTING, evt 0x0
[   876][D][BluetoothA2DPSource.cpp:632] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks
[   886][D][BluetoothA2DPSource.cpp:641] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks ESP_A2D_CONNECTION_STATE_EVT: Connecting
[   899][D][BluetoothA2DPSource.cpp:722] bt_app_av_state_connecting_hdlr(): [BT_AV] bt_app_av_state_connecting_hdlr evt 0
[  4226][D][BluetoothA2DPSource.cpp:461] bt_app_gap_callback(): [BT_AV] bt_app_gap_callback evt 16
[  4234][I][BluetoothA2DPSource.cpp:545] bt_app_gap_callback(): [BT_AV] ESP_BT_GAP_ACL_CONN_CMPL_STAT_EVT
[  4585][D][BluetoothA2DPSource.cpp:257] bt_app_work_dispatch(): [BT_API] bt_app_work_dispatch event 0x0, param len 16
[  4586][D][BluetoothA2DPSource.cpp:307] bt_app_task_handler(): [BT_API] bt_app_task_handler, signal: 0x1, event: 0x0
[  4595][I][BluetoothA2DPSource.cpp:665] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state APP_AV_STATE_CONNECTING, evt 0x0
[  4607][D][BluetoothA2DPSource.cpp:632] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks
[  4616][D][BluetoothA2DPSource.cpp:641] process_user_state_callbacks(): [BT_AV] process_user_state_callbacks ESP_A2D_CONNECTION_STATE_EVT: Connected
[  4629][D][BluetoothA2DPSource.cpp:722] bt_app_av_state_connecting_hdlr(): [BT_AV] bt_app_av_state_connecting_hdlr evt 0
[  4640][I][BluetoothA2DPSource.cpp:730] bt_app_av_state_connecting_hdlr(): [BT_AV] a2dp connected

**Comment by pschatzmann** [comment]
My Bluetooth speaker automatically enters into paring mode when switched on and is therefore discoverable (and it announces itself).
In your case you see in the log that the ESP32 is scanning for devices but no device is announcing itself, so nothing is found and therefore there is no connection!

So I confirm that the speaker must be in pairing mode (or at least in Discoverable mode) to be found and if you want to avoid this you would need to use the auto_reconnect = true: so that is automatically connects to the last device.

I have only this one Bluetooth speaker and using the ESP32 as Bluetooth Sink will be be behaving the same way, so no I was not aware of this.

**Comment by Prog32aw** [comment]
If someone wants to use this library in an Bluetooth Source project it does not work as users will expect. Nobody wants to put devices in pairing mode to simply connect them (is is sometimes not intuitive but complicated and the only reason to read the documentation). And ist is not unusual, that users have more than one sink paired with a source.

Is it - without modifying the A2DPBluetoothSource library - possible to do the logic of maintaining several devices with name / address and connect them by function calls? I took a look at the sources and found connect_to(addr), which possible could do this. In a2dp_source.start(...) one paramater is the data request callback reference. How is this callback passed to the library when using connect_to(addr)?

Thanks

**Comment by pschatzmann** [comment]
Please read the Design Overivew Wiki which describes how to extend the functionality w/o modifying the source code.
Just check the logic related to reconnect_status to get you started.
